docs/proposals/failover/application-failover.md

```diff
@@ -0,0 +1,562 @@
+Therefore, we need a parameter to tolerate the unhealthy application.
+
+### API change
+
+#### PropagationPolicy API change
+
+```golang
+// PropagationSpec represents the desired behavior of PropagationPolicy.
+type PropagationSpec struct {
+	// ....
+
+	PropagateDeps bool `json:"propagateDeps,omitempty"`
+
+	// Failover indicates how Karmada migrates applications in case of failures.
+	// If this value is nil, failover is disabled.
+	// +optional
+	Failover *FailoverBehavior `json:"failover,omitempty"`
+
+	// ...
+}
+
+// FailoverBehavior indicates failover behaviors in case of an application or
+// cluster failure.
+type FailoverBehavior struct {
+    // Application indicates failover behaviors in case of application failure.
+    // If this value is nil, failover is disabled.
+    // If set, the PropagateDeps should be true so that the dependencies could
+    // be migrated along with the application.
+    // +optional
+    Application *ApplicationFailoverBehavior `json:"application,omitempty"`
+  
+    // Cluster indicates failover behaviors in case of cluster failure.
+    // If this value is nil, failover is disabled.
+    // +optional
+    Cluster *ClusterFailoverBehavior `json:"cluster,omitempty"`
+}
+
+// ApplicationFailoverBehavior indicates application failover behaviors.
+type ApplicationFailoverBehavior struct {
+    // InitiateGates indicates the preconditions of the failover process.
+    // If specified, all gates will be evaluated for prerequisites.
+    // +optional
+    InitiateGates *FailoverInitiateGate `json:"initiateGates,omitempty"`
+  
+    // ActiveGates indicates the conditions of performing the failover process.
+    //
+    // +kubebuilder:validation:MinItems=1
+    // +required
+    ActiveGates *FailoverActiveGate `json:"activeGates"`
+  
+    // PurgeMode represents how to deal with the legacy applications on the
+    // cluster from which the application is migrated.
+    // Valid options are "Immediately", "Graciously" and "Never".
+    // Defaults to "Graciously".
+    //
+    // +kubebuilder:validation:Enum=Immediately;Graciously;Never
+    // +optional
+    PurgeMode string `json:"purgeMode,omitempty"`
+  
+    // EscapeSeconds represents the period of time the cluster from which the
+    // application was migrated from should be escaped.
+    // Defaults to 600s.
+    //
+    // +kubebuilder:default=600
+    // +optional
+    EscapeSeconds int32 `json:"escapeSeconds,omitempty"`
+}
+
+
+// FailoverInitiateGate represents the preconditions of the failover process.
+type FailoverInitiateGate struct {
+    // DelaySeconds refers to a period of time after the application has started before
+    // the application failover is initiated.
+    // +optional
+    DelaySeconds int32 `json:"delaySeconds,omitempty"`
+  
+    // HealthyState refers to the healthy status reported by the Karmada resource
+    // interpreter.
```

Comment:
I think there is no need to specifically point out the resource interpreter in this comment.

Reply:
I think filling it in allows users to understand how this field is configured.